pkg/controller/controller.go

```diff
@@ -0,0 +1,266 @@
+
+	// If set to any other value than 0, will cause DoFunc to be run in the
+	// specified interval. The interval starts from when the DoFunc has
+	// returned last
+	RunInterval time.Duration
+
+	// ErrorRetryBaseDuration is the initial time to wait to run DoFunc
+	// again on return of an error. On each consecutive error, this value
+	// is multiplied by the number of consecutive errors to provide a
+	// constant back off. The default is 1s.
+	ErrorRetryBaseDuration time.Duration
+
+	// NoErrorRetry when set to true, disabled retries on errors
+	NoErrorRetry bool
+
+	successCount   int
+	failureCount   int
+	lastError      error
+	lastErrorStamp time.Time
+	uuid           string
+	stop           chan struct{}
+}
+
+// GetSuccessCount returns the number of successful controller runs
+func (c *Controller) GetSuccessCount() int {
+	c.RLock()
+	defer c.RUnlock()
+
+	return c.successCount
+}
+
+// GetFailureCount returns the number of failed controller runs
+func (c *Controller) GetFailureCount() int {
+	c.RLock()
+	defer c.RUnlock()
+
+	return c.failureCount
+}
+
+// GetLastError returns the last error returned
+func (c *Controller) GetLastError() error {
+	c.RLock()
+	defer c.RUnlock()
+
+	return c.lastError
+}
+
+// GetLastErrorTimestamp returns the last error returned
+func (c *Controller) GetLastErrorTimestamp() time.Time {
+	c.RLock()
+	defer c.RUnlock()
+
+	return c.lastErrorStamp
+}
+
+func (c *Controller) runController() {
+	errorRetries := 1
+
+	for {
+		var (
+			err      error
+			interval = c.RunInterval
+		)
+
+		if c.DoFunc != nil {
+			err = c.DoFunc()
+		} else {
+			err = fmt.Errorf("DoFunc is nil")
+		}
+
+		if err != nil {
+			log.WithFields(logrus.Fields{
+				fieldControllerName:    c.Name,
+				fieldUUID:              c.uuid,
+				fieldConsecutiveErrors: errorRetries,
+			}).WithError(err).Debug("Controller run failed")
+
+			c.Lock()
+			c.lastError = err
+			c.lastErrorStamp = time.Now()
+			c.failureCount++
+			c.Unlock()
+
+			if !c.NoErrorRetry {
+				if c.ErrorRetryBaseDuration != time.Duration(0) {
+					interval = time.Duration(errorRetries) * c.ErrorRetryBaseDuration
+				} else {
+					interval = time.Duration(errorRetries) * time.Second
+				}
+
+				errorRetries++
+			}
+		} else {
+			c.Lock()
+			c.successCount++
+			c.Unlock()
+
+			// reset error retries after successful attempt
+			errorRetries = 1
+
+			// If no run interval is specified, no further updates
+			// are required and we can shut down the controller
+			if c.RunInterval == time.Duration(0) {
+				goto shutdown
+			}
+		}
+
+		select {
+		case <-c.stop:
+			goto shutdown
+
+		case <-time.After(interval):
+		}
+
+	}
+
+shutdown:
+	log.WithFields(logrus.Fields{
+		fieldControllerName: c.Name,
+		fieldUUID:           c.uuid,
+	}).Debug("Shutting down controller")
+}
+
+func (c *Controller) stopController() {
+	close(c.stop)
+}
+
+type controllerMap map[string]*Controller
+
+// Manager is a list of controllers
+type Manager struct {
+	lock.RWMutex
```

Comment:
add a name to this attribute otherwise the `Lock()` and `Unlock()` are public.

Reply:
fixed